babel/number_spelling/__init__.py

```diff
@@ -0,0 +1,551 @@
+    """
+    Represents a number and its internal structure.
+
+    The number is separated into integer and fraction parts
+    and the fraction part is also rounded if necessary.
+    """
+    def __init__(self, value, integer, fraction, rounded, negative):
+        super(NumberContext, self).__init__()
+        self.value = value
+        self.integer = SideContext(integer, self)
+        self.fraction = SideContext(fraction, self)
+        self.rounded = rounded
+        self.negative = negative
+
+    def __iter__(self):
+        yield self.fraction
+        yield self.integer
+
+    def __len__(self):
+        return len(self.integer) + len(self.fraction)
+
+    @property
+    def last_nonzero(self):
+        """
+        The first non-zero digit from the right
+        """
+        for side in self:
+            for group in side:
+                for digit in group:
+                    if digit.value != 0:
+                        return digit
+
+
+class SideContext(ContextBase):
+    """
+    Represent one side of the decimal point
+    """
+    def __init__(self, groups, number):
+        super(SideContext, self).__init__()
+        self.number = number
+        self.groups = tuple(GroupContext(digits, number, self, i) for i, digits in enumerate(groups))
+        # calculate the value of the side the hard way (might use number.value instead)
+        e, s = 0, 0
+        for g in self.groups:
+            s += g.value*10**e
+            e += len(g)
+        self.value = s
+
+    def __iter__(self):
+        for group in self.groups:
+            yield group
+
+    def __len__(self):
+        return sum(len(g) for g in self)
+
+
+class GroupContext(ContextBase):
+    """
+    Represent a group of digits inside a number
+
+    Digits inside the group are indexed from the right.
+    """
+    def __init__(self, digits, number, side, index):
+        super(GroupContext, self).__init__()
+        self.digits = tuple(DigitContext(v, number, side, self, i) for i,v in enumerate(digits))
+        self.number = number
+        self.side = side
+        self.index = index
+        self.value = sum(d.value*dec(10)**e for e,d in enumerate(self.digits)) # might use int
+
+    def __len__(self):
+        return len(self.digits)
+
+    def __iter__(self):
+        for digit in self.digits:
+            yield digit
+
+
+class DigitContext(ContextBase):
+    """
+    Represent one digit of a number and its context
+    """
+    def __init__(self, value, number, side, group, index):
+        super(DigitContext, self).__init__()
+        self.value = value
+        self.number = number
+        self.side = side
+        self.group = group
+        self.index = index
+        self.string = ''
+
+    def __iter__(self):
+        """
+        Not so nice, but this way it could also inherit from ContextBase
+        """
+        yield self.string
+
+
+#####################################
+# locale specific speller functions #
```

Comment:
@erickwilder and @sils1297 as I mentioned earlier this structure could work only if I import all the submodules when the ``__init__.py`` runs, otherwise locale negotiation would not work. It this case the import mechanism gives a little overhead to the module, but the structure improves. I was thinking about a more elegant solution, but couldn't come up with one yet.

(We should make some connection with the CLDR here, but the number spelling in CLDR is seriously broken. It could be used for flagging which locale has a spelling implementation or we should come up with an internal flagging. Both ways the import mechanism could become more dynamic and thus efficient.)